src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/BuildAssetsUtils.cs

```diff
@@ -22,428 +22,693 @@ namespace NuGet.Commands
+                                                                repositories,
+                                                                sortedPkg.Id,
+                                                                sortedPkg.Version))))
+                                                .Where(e => e.Key != null)
+                                                .ToArray();
+
+                // build/ {packageId}.targets
+                var buildTargetsGroup = new MSBuildRestoreItemGroup();
+                buildTargetsGroup.RootName = MSBuildRestoreItemGroup.ImportGroup;
+                buildTargetsGroup.Position = 2;
+
+                buildTargetsGroup.Items.AddRange(sortedPackages.SelectMany(pkg =>
+                    pkg.Key.Build.WithExtension(TargetsExtension)
+                        .Where(e => pkg.Value.Exists())
+                        .Select(e => pkg.Value.GetAbsolutePath(e)))
+                        .Select(path => GetImportPath(path, repositoryRoot))
+                        .Select(GenerateImport));
+
+                targets.AddRange(GetGroupsWithConditions(buildTargetsGroup, isMultiTargeting, frameworkConditions));
+
+                // props/ {packageId}.props
+                var buildPropsGroup = new MSBuildRestoreItemGroup();
+                buildPropsGroup.RootName = MSBuildRestoreItemGroup.ImportGroup;
+                buildPropsGroup.Position = 2;
+
+                buildPropsGroup.Items.AddRange(sortedPackages.SelectMany(pkg =>
+                    pkg.Key.Build.WithExtension(PropsExtension)
+                        .Where(e => pkg.Value.Exists())
+                        .Select(e => pkg.Value.GetAbsolutePath(e)))
+                        .Select(path => GetImportPath(path, repositoryRoot))
+                        .Select(GenerateImport));
+
+                props.AddRange(GetGroupsWithConditions(buildPropsGroup, isMultiTargeting, frameworkConditions));
+
+                if (isMultiTargeting)
                 {
-                    includeLibrary = libraryFlags.HasFlag(LibraryIncludeFlags.Build);
+                    // buildMultiTargeting/ {packageId}.targets
+                    var buildCrossTargetsGroup = new MSBuildRestoreItemGroup();
+                    buildCrossTargetsGroup.RootName = MSBuildRestoreItemGroup.ImportGroup;
+                    buildCrossTargetsGroup.Position = 0;
+
+                    buildCrossTargetsGroup.Items.AddRange(sortedPackages.SelectMany(pkg =>
+                        pkg.Key.BuildMultiTargeting.WithExtension(TargetsExtension)
+                            .Where(e => pkg.Value.Exists())
+                            .Select(e => pkg.Value.GetAbsolutePath(e)))
+                            .Select(path => GetImportPath(path, repositoryRoot))
+                            .Select(GenerateImport));
+
+                    targets.AddRange(GetGroupsWithConditions(buildCrossTargetsGroup, isMultiTargeting, CrossTargetingCondition));
```

Comment:
Forcing the TFM conditions also avoid scenarios where the user adds a new TFM to $(TargetFrameworks) but hasn't done a restore yet.